sdk/core/test/integration/otp/index.test.ts

```diff
@@ -1,417 +1,426 @@
 import 'mocha'
-
 import '../env'
 
 import { expect } from 'chai'
+import cryptoRandomString from 'crypto-random-string'
+
 import { CommonNetworkMember } from '../../../src/CommonNetworkMember'
 import { SdkOptions } from '../../../src/dto/shared.dto'
+import SdkError from '../../../src/shared/SdkError'
 
-import { getOtp, generateUsername, generateEmail, getOptionsForEnvironment } from '../../helpers'
+import { generateUsername, getOptionsForEnvironment } from '../../helpers'
 import { MessageParameters } from '../../../dist/dto'
-import { TestEmailHelper } from '../../helpers/getOtpFromTestMailApp'
+import { TestmailHelper } from '../../helpers/TestmailHelper'
 
-const { TEST_SECRETS } = process.env
-const { COGNITO_PASSWORD } = JSON.parse(TEST_SECRETS)
+const { COGNITO_PASSWORD } = JSON.parse(process.env.TEST_SECRETS)
 
 const options: SdkOptions = getOptionsForEnvironment()
+const { env } = getOptionsForEnvironment(true)
+
+const prepareOtpMessageParameters = (
+  testId: string,
+  { suffix }: { suffix?: string } = {},
+): [string, string, MessageParameters] => {
+  const tag = generateTag(testId, suffix)
+  const email = TestmailHelper.generateEmailForTag(tag)
+  const messageParameters: MessageParameters = {
+    message: `Your verification code is: {{CODE}}`,
+    subject: `Verification code`,
+  }
+
+  return [email, tag, messageParameters]
+}
+
+const waitForOtpCode = async (tag: string, timestampFrom?: number): Promise<string> => {
+  const { text, html } = await TestmailHelper.waitForNewEmail(tag, timestampFrom)
```

Comment:
`livequery` immediately returns the result if there is at least one matching email in the box, so your proposed logic might look too complicated in the actual code
I think we should concern ourselves with such questions only if we have concrete problems to solve.
If clock is faulty, then there are more problems then just integration tests failing - which we will see immediately.

Reply:
> livequery immediately returns the result if there is at least one matching email in the box

Which is why I said "call api with livequery=true **and timestamp=lastMessageTimestamp**" :)

> If clock is faulty, then there are more problems then just integration tests failing

It does not have to be faulty, it just has to be a couple of seconds behind or ahead of clock on their server. It happens all the time, that's why we have NTP to keep it in check :)
Generally, relying on the fact that your clock is in perfect sync with remote clock is quite a bad idea; it is not guaranteed to be.
And there is a way to avoid relying on it (by supplying testmail with timestamps they generate, instead of our timestamps).

Reply:
Yeah, I agree, we can utilize their timestamps :+1:

Reply:
(In the context of another discussion there) I would suggest to do something along the lines of:
```
class Mailbox {
  readonly email: string
  private readonly tag: string
  private lastMessageTimestamp = 0

  Mailbox() {
    this.tag = generateTag()
    this.email = generateEmailFromTag()
  }

  fetchNewMessage() {
    const url = new URL(TESTMAIL_API_URL)
    url.searchParams = new URLSearchParams({
      tag: this.tag,
      timestamp_from: this.lastMessageTimestamp,
      livequery: 'true',
      ...
    )
    const response = await fetch(url)
    const { emails } = await response.json()
    const { timestamp, subject, html, text } = emails[0]
    this.lastMessageTimestamp = timestamp
    return { subject, html, text }
  }
}
```

That way we would not have to parse email addresses for tags.